I/O Timeout

Howdy folks,

No matter running the latest version or not, or latest Debian (9) or not, on RAID 10, with 30 transfers and full verbose, we get:

ERRO[0124] Unable to create object part.                 cause=read tcp 46.4.*.*:443->89.114.13*.*:49244: i/o timeout source=[object-handlers.go:817:objectAPIHandlers.PutObjectPartHandler()] stack=fs-v1-helpers.go:272:fsCreateFile fs-v1-multipart.go:523:fsObjects.PutObjectPart <autogenerated>:339:(*fsObjects).PutObjectPart object-handlers.go:814:objectAPIHandlers.PutObjectPartHandler api-router.go:46:(objectAPIHandlers).PutObjectPartHandler-fm


Any hint?

Disks are fine, plenty of IO available. And these are mostly EPS files, jpg, etc, only 63GB total. I got this very same error in two servers.

@harshavardhana, I am using mounted folders via -v. Docker using aufs driver. 
When using Java SDK and `MinioClient.setTimeout(0, 0, 0)` the Java hangs forever.
Otherwise it timeouts with 
```java
java.net.SocketTimeoutException: timeout
	at okio.Okio$4.newTimeoutException(Okio.java:227)
	at okio.AsyncTimeout.exit(AsyncTimeout.java:284)
	at okio.AsyncTimeout$1.write(AsyncTimeout.java:183)
	at okio.RealBufferedSink.emitCompleteSegments(RealBufferedSink.java:171)
	at okio.RealBufferedSink.write(RealBufferedSink.java:41)
	at okhttp3.internal.http1.Http1Codec$FixedLengthSink.write(Http1Codec.java:286)
	at okio.RealBufferedSink.emitCompleteSegments(RealBufferedSink.java:171)
	at okio.RealBufferedSink.write(RealBufferedSink.java:109)
	at io.minio.MinioClient$3.writeTo(MinioClient.java:871)
	at okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.java:62)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92)
	at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:45)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67)
	at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:93)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67)
	at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92)
	at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:120)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67)
	at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:185)
	at okhttp3.RealCall.execute(RealCall.java:69)
	at io.minio.MinioClient.execute(MinioClient.java:998)
	at io.minio.MinioClient.executePut(MinioClient.java:1266)
	at io.minio.MinioClient.executePut(MinioClient.java:1288)
	at io.minio.MinioClient.putObject(MinioClient.java:3480)
	at io.minio.MinioClient.putObject(MinioClient.java:3590)
	at io.minio.MinioClient.putObject(MinioClient.java:3107)
	at test.TestMinio.test(TestMinio.java:56)
Caused by: java.net.SocketException: Socket closed
	at java.net.SocketOutputStream.socketWrite(SocketOutputStream.java:116)
	at java.net.SocketOutputStream.write(SocketOutputStream.java:153)
	at okio.Okio$1.write(Okio.java:78)
	at okio.AsyncTimeout$1.write(AsyncTimeout.java:179)
	... 52 more
```

and this entry at minio-server log:
```bash
time="2017-09-04T04:25:12Z" level=error msg="Unable to create object part." cause="#has incomplete body" source="[object-handlers.go:870:objectAPIHandlers.PutObjectPartHandler()]" stack="fs-v1-helpers.go:275:fsCreateFile fs-v1-multipart.go:555:fsObjects.PutObjectPart <autogenerated>:402:(*fsObjects).PutObjectPart object-handlers.go:867:objectAPIHandlers.PutObjectPartHandler api-router.go:46:(objectAPIHandlers).PutObjectPartHandler-fm"
```

--- a/cmd/globals.go
+++ b/cmd/globals.go
@@ -64,8 +64,13 @@ const (
 	// Limit memory allocation to store multipart data
 	maxFormMemory = int64(5 * humanize.MiByte)
 
-	// The maximum allowed difference between the request generation time and the server processing time
-	globalMaxSkewTime = 15 * time.Minute
+	// The maximum allowed time difference between the incoming request
+	// date and server date during signature verification.
+	globalMaxSkewTime = 15 * time.Minute // 15 minutes skew allowed.
+
+	// Default Read/Write timeouts for each connection.
+	globalConnReadTimeout  = 15 * time.Minute // Timeout after 15 minutes of no data sent by the client.
+	globalConnWriteTimeout = 15 * time.Minute // Timeout after 15 minutes if no data received by the client.
 )
 
 var (
@@ -141,12 +146,9 @@ var (
 	globalObjectTimeout    = newDynamicTimeout( /*1*/ 10*time.Minute /*10*/, 600*time.Second)  // timeout for Object API related ops
 	globalOperationTimeout = newDynamicTimeout(10*time.Minute /*30*/, 600*time.Second)         // default timeout for general ops
 	globalHealingTimeout   = newDynamicTimeout(30*time.Minute /*1*/, 30*time.Minute)           // timeout for healing related ops
-)
 
-var (
-	// Keeps the connection active by waiting for following amount of time.
-	// Primarily used in ListenBucketNotification.
-	globalSNSConnAlive = 5 * time.Second
+	// Keep connection active for clients actively using ListenBucketNotification.
+	globalSNSConnAlive = 5 * time.Second // Send a whitespace every 5 seconds.
 )
 
 // global colors.

--- a/cmd/server-main.go
+++ b/cmd/server-main.go
@@ -193,6 +193,8 @@ func serverMain(ctx *cli.Context) {
 	initGlobalAdminPeers(globalEndpoints)
 
 	globalHTTPServer = miniohttp.NewServer([]string{globalMinioAddr}, handler, globalTLSCertificate)
+	globalHTTPServer.ReadTimeout = globalConnReadTimeout
+	globalHTTPServer.WriteTimeout = globalConnWriteTimeout
 	globalHTTPServer.UpdateBytesReadFunc = globalConnStats.incInputBytes
 	globalHTTPServer.UpdateBytesWrittenFunc = globalConnStats.incOutputBytes
 	globalHTTPServer.ErrorLogFunc = errorIf

